ott/core/gw_barycenter.py

```diff
@@ -0,0 +1,323 @@
+        geom_yy = geometry.Geometry(y, epsilon=eps, scale_cost=scale)
+      else:
+        geom_yy = pointcloud.PointCloud(
+            y, cost_fn=cost_fn, epsilon=eps, scale_cost=scale
+        )
+      if problem.is_fused:
+        geom_xy = pointcloud.PointCloud(
+            state.x, f, cost_fn=cost_fn, epsilon=eps, scale_cost=scale
+        )
+      else:
+        geom_xy = None
+
+      quad_problem = quad_problems.QuadraticProblem(
+          geom_xx=geom_xx,
+          geom_yy=geom_yy,
+          geom_xy=geom_xy,
+          a=state.a,
+          b=b,
+          fused_penalty=problem.fused_penalty,
+      )
+      out = self._quad_solver(quad_problem)
+
+      return (
+          out.reg_gw_cost, out.convergence, out.matrix,
+          out.errors if store_errors else None
+      )
+
+    in_axes = [None, 0, 0]
+    in_axes += [0] if problem.is_fused else [None]
+    solve_fn = jax.vmap(solve_gw, in_axes=in_axes)
+
+    y, b = problem.segmented_y_b
+    y_f = problem.segmented_y_fused
+    costs, convergeds, transports, errors = solve_fn(state, b, y, y_f)
+
+    cost = jnp.sum(costs * problem.weights)
+    costs = state.costs.at[iteration].set(cost)
+    converged = jnp.all(convergeds)
+    gw_convergence = state.gw_convergence.at[iteration].set(converged)
+
+    if self.store_inner_errors:
+      errors = state.errors.at[iteration, ...].set(errors)
+    else:
+      errors = None
+
+    x_new = problem.update_features(transports, state.a)
+    c_new = problem.update_barycenter(transports, state.a)
+    return state.set(
+        x=x_new,
+        c=c_new,
+        costs=costs,
+        errors=errors,
+        gw_convergence=gw_convergence
+    )
+
+  def output_from_state(self, state: GWBarycenterState) -> GWBarycenterState:
+    """No-op."""
+    # TODO(michalk8): just for consistency with continuous barycenter
+    # will be refactored in the future
+    return state
+
+  def tree_flatten(self) -> Tuple[Sequence[Any], Dict[str, Any]]:
+    children, aux = super().tree_flatten()
+    aux["quad_solver"] = self._quad_solver
+    return children, aux
+
+  @classmethod
+  def tree_unflatten(
+      cls, aux_data: Dict[str, Any], children: Sequence[Any]
+  ) -> "GromovWassersteinBarycenter":
+    epsilon, _, _, threshold = children
+    return cls(
+        epsilon=epsilon,
+        threshold=threshold,
+        **aux_data,
+    )
+
+
+@partial(jax.vmap, in_axes=[None, 0, None, 0, None])
+def init_transports(
+    solver, key: jnp.ndarray, a: jnp.ndarray, b: jnp.ndarray,
+    epsilon: Optional[float]
+) -> jnp.ndarray:
+  """Initialize random cost matrix and solve the OT problem.
+
+  Args:
+    solver: Linear OT solver.
+    key: Random key.
+    a: Source marginals (e.g., for barycenter) of shape ``[B,]``.
+    b: Target marginals of shape ``[N,]``.
+    epsilon: Entropy regularization.
+
+  Returns:
+    Transport map of shape ``[B, N]``.
+  """
+  cost = jax.random.uniform(key, shape=(len(a), len(b)), minval=0, maxval=1)
```

Comment:
intuitively, i guess it might make sense to initialize with something that looks indeed like a cost matrix, at least symmetric. maybe a random pointcloud in arbitrary dimensions (e.g. 2d?)

Reply:
Nice catch, was following the original impl. for consistency. Now sample a random PC in 2D.